### PR TITLE
build: single, optimized Dockerfile for all services

### PR DESCRIPTION
**Pull request by lklimek**
## Issue being fixed or feature implemented

Docker build process is split into several Dockerfiles. This is quite hard to manage (need to replicate any improvements between Dockerfiles) and not optimal (some parts of Dockerfile doing the same logic are executed multiple times).

## What was done?

* Merged all /packages/*/Dockerfile (except Envoy) into one /Dockerfile
* Optimized for sccache and cache mounts
* Build javascript artifacts inside docker
* Added documentation
* Updated dashmate and github actions accordingly

## How Has This Been Tested?

* locally, with `time docker buildx build -f Dockerfile --build-arg SCCACHE_MEMCACHED=tcp://172.17.0.1:11211 --progress=plain .`
* in local multipass VM, using `yarn start`
* inside github actions

## Breaking Changes

Changed Dockerfile locations and structure can have impact on people building Docker images outside of dashmate.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -1974,7 +1974,7 @@ dependencies = [
  "regex",
  "serde",
  "serde_json",
- "time 0.3.20",
```
Comment (shumkov) [inline Cargo.lock:1977]: Do we need these updates for this PR?

File: Dockerfile [hunk 2]
```diff
@@ -64,23 +71,26 @@ RUN if [[ "$TARGETARCH" == "arm64" ]] ; then export SCC_ARCH=aarch64; else expor
...
+#         /opt/emsdk/emsdk activate latest
 
-RUN curl -Ls \
-        https://github.com/emscripten-core/emsdk/archive/refs/tags/3.1.36.tar.gz | \
-        tar -C /opt -xz && \
-        ln -s /opt/emsdk-* /opt/emsdk && \
-        /opt/emsdk/emsdk install latest && \
-        /opt/emsdk/emsdk activate latest
+# Configure Node.js
+RUN npm install -g npm@latest && \
```
Comment (shumkov) [inline Dockerfile:83]: I would set specific version (latest 16)

File: Dockerfile [hunk 3]
```diff
@@ -64,23 +71,26 @@ RUN if [[ "$TARGETARCH" == "arm64" ]] ; then export SCC_ARCH=aarch64; else expor
...
+    npm install -g corepack@latest && \
+    corepack prepare yarn@stable --activate && \
```
Comment (shumkov) [inline Dockerfile:85]: yarn@3.3.0

File: Dockerfile [hunk 4]
```diff
@@ -94,62 +104,107 @@ ARG ACTIONS_RUNTIME_TOKEN
 ARG SCCACHE_MEMCACHED
 
 # Disable incremental buildings, not supported by sccache
-ENV CARGO_INCREMENTAL=false
+ARG CARGO_INCREMENTAL=false
 
 # Select whether we want dev or release
-ARG CARGO_BUILD_PROFILE=dev
+ARG CARGO_BUILD_PROFILE=debug
```
Comment (shumkov) [inline Dockerfile:110]: It seems there is no profile `debug`. We should use `dev` instead and fix scripts where we expect `CARGO_BUILD_PROFILE == debug`

File: Dockerfile [hunk 5]
```diff
@@ -94,62 +104,107 @@ ARG ACTIONS_RUNTIME_TOKEN
...
+# Install wasm-bindgen-cli
+WORKDIR /platform
+RUN --mount=type=cache,sharing=shared,target=/root/.cache/sccache \
+    --mount=type=cache,sharing=shared,target=${CARGO_HOME}/registry/index \
+    --mount=type=cache,sharing=shared,target=${CARGO_HOME}/registry/cache \
+    --mount=type=cache,sharing=shared,target=${CARGO_HOME}/git/db \
+    --mount=type=cache,sharing=shared,target=/platform/target \
+    export SCCACHE_SERVER_PORT=$((RANDOM+1025)) && \
+    if [[ -z "${SCCACHE_MEMCACHED}" ]] ; then unset SCCACHE_MEMCACHED ; fi ; \
+    cargo install wasm-bindgen-cli
```
Comment (shumkov) [inline Dockerfile:125]: So we go with both sccache and target?

Comment (strophy) [reply Dockerfile:125]: I thought we agreed not to use sccache?

File: Dockerfile [hunk 6]
```diff
@@ -29,28 +30,35 @@ FROM rust:alpine${ALPINE_VERSION} as deps
...
+        binutils \
+        ca-certificates \        
         clang-static clang-dev \
-        llvm-static llvm-dev  \
+        cmake \
         git \
+        libc-dev \        
         linux-headers \
+        llvm-static llvm-dev  \
+        nodejs \
```
Comment (shumkov) [inline Dockerfile:41]: we need to install 16.x

File: Dockerfile [hunk 7]
```diff
@@ -256,3 +310,112 @@ ENTRYPOINT ["/usr/bin/drive-abci"]
...
 EXPOSE 26658
+
+#
+# STAGE: DASHMATE BUILD
+#
+FROM build-wasm-dpp AS build-dashmate
+
+# Install Test Suite specific dependencies using previous
+# node_modules directory to reuse built binaries
+RUN --mount=type=cache,target=/tmp/unplugged \
```
Comment (shumkov) [inline Dockerfile:321]: Are we building JS packages somewhere?

File: Dockerfile (hunk 6 above)
Comment (strophy) [reply Dockerfile:41]: On Alpine 3.16, this will always install 16.x

File: packages/rs-dpp/src/block/block_info.rs [hunk 8]
```diff
@@ -7,7 +7,7 @@ use bincode::{Decode, Encode};
 use platform_serialization::{PlatformDeserialize, PlatformSerialize};
```
Comment (shumkov) [inline packages/rs-dpp/src/block/block_info.rs:7]: PR contains a lot of changes are not related to the subject

**Review by shumkov — CHANGES_REQUESTED** [review]

**Review by strophy — APPROVED** [review]
Thanks, this makes the build process much clearer to understand

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+FROM node:16-alpine${ALPINE_VERSION} AS test-suite
+
+RUN apk add --no-cache bash
+
+LABEL maintainer="Dash Developers <dev@dash.org>"
+LABEL description="Dash Platform test suite"
+
+WORKDIR /platform
+
+COPY --from=build-test-suite /platform /platform
```
Comment (shumkov) [inline Dockerfile:299]: Shouldn't we copy only necessary dependencies?

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+WORKDIR /var/lib/dash/rs-drive-abci
+ENTRYPOINT ["/usr/bin/drive-abci"]
+CMD ["-vvvv", "start"]
+
+EXPOSE 26658
+
+#
+# STAGE: DASHMATE BUILD
+#
+FROM build-js AS build-dashmate
```
Comment (shumkov) [inline Dockerfile:231]: I was a bit confused about what dashmate image we have since we don't dockerize dashmate. It took me some time to figure out that we have dashmate-helper subcomponent that we run together with dashmate services. What do you think @strophy @lklimek should we rename it to dashmate-helper?

File: README.md [hunk 11]
```diff
@@ -53,20 +53,24 @@ this repository may be used on the following networks:
...
   - [rust](https://www.rust-lang.org/tools/install) v1.67+
-  - [wasm-bingen toolchain](https://rustwasm.github.io/wasm-bindgen/): 
+  - [wasm-bingen toolchain](https://rustwasm.github.io/wasm-bindgen/):
     - **IMPORTANT (OSX only)**: built-in `llvm` on OSX does not work, needs to be installed from brew:
       - `brew install llvm`
       - LLVM installed from brew is keg only, and path to it must be provided in the profile file, e.g.`echo 'export PATH="/opt/homebrew/opt/llvm/bin:$PATH"' >> ~/.zshrc`
-    - `cargo install wasm-bindgen-cli` 
+    - install `protoc` - protobuf compiler:
+      - on debian/ubuntu: `apt install -y protobuf-compiler`
+      - on other systems, install most recent version from [Protocol Buffers releases page](https://github.com/protocolbuffers/protobuf/releases) (tested with protobuf 22.4)
```
Comment (shumkov) [inline README.md:62]: Let's add `brew install protobuf` for Mac since the majority of DCG devs use it

File: packages/dashmate/docker-compose.platform.build.yml [hunk 12]
```diff
@@ -4,7 +4,8 @@ services:
   drive_abci:
     build:
       context: ${PLATFORM_SOURCE_PATH:?err}
-      dockerfile: ${PLATFORM_SOURCE_PATH:?err}/packages/rs-drive-abci/Dockerfile
+      dockerfile: ${PLATFORM_SOURCE_PATH:?err}/Dockerfile
+      target: drive-abci
       args:
         SCCACHE_MEMCACHED: ${SCCACHE_MEMCACHED}
```
Comment (shumkov) [inline packages/dashmate/docker-compose.platform.build.yml:10]: Cool, this is much better! Let's rip out https://github.com/dashpay/platform/blob/15b3ac912d34736ef4342db46280dddfe66ab29c/packages/dashmate/src/listr/tasks/buildServicesTaskFactory.js#L23

File: packages/dashmate/docker/entrypoint.sh [hunk 13]
```diff
@@ -5,30 +5,37 @@ DOCKER_GROUP_ID=$(stat -c %g /var/run/docker.sock)
 USERNAME=dashmate
 GROUP=docker
 
-# check if user with our uid exists in the system
```
Comment (shumkov) [inline packages/dashmate/docker/entrypoint.sh:8]: Had you changed only indentation? We have a PR where guys are fixing logic there https://github.com/dashpay/platform/pull/1072

File: packages/wasm-dpp/scripts/build-wasm.sh [hunk 14]
```diff
@@ -1,46 +1,43 @@
...
+if ! [[ -d ${OUTPUT_DIR} ]]; then
+  mkdir -p "${OUTPUT_DIR}"
 fi
 
-if ! [ -x "$(command -v wasm-bindgen)" ]; then
-    echo 'Wasm-bindgen CLI is not installed. Installing';
-    cargo install -f wasm-bindgen-cli
+if ! [[ -x "$(command -v wasm-bindgen)" ]]; then
+  echo 'Wasm-bindgen CLI is not installed. Installing'
+  cargo install --config net.git-fetch-with-cli=true -f wasm-bindgen-cli
```
Comment (shumkov) [inline packages/wasm-dpp/scripts/build-wasm.sh:28]: Shall we install specific version here as well?

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/dashmate/docker/entrypoint.sh (hunk 13 above)
Comment (shumkov) [reply packages/dashmate/docker/entrypoint.sh:8]: Agree!

File: Dockerfile (hunk 10 above)
Comment (shumkov) [reply Dockerfile:228]: We discussed it on a call and agreed that it makes sense to rename

**Review by shumkov — APPROVED** [review]
👍

**Review by strophy — APPROVED** [review]
